plugin/cache: Add refresh mode setting to serve_stale

atollena opened the pull request:
### 1. Why is this pull request needed and what does it do?

This PR adds an optional REFRESH_MODE parameter on the serve_stale configuration directive of the cache plugin. For context, please see #4309

### 2. Which issues (if any) are related?

Reasoning and RFC discussed at #4309
Other similar issues are #4097 and #4199.

### 3. Which documentation changes (if any) need to be made?

I updated the cache plugin README.md to document the new parameter on the PR.

### 4. Does this introduce a backward incompatible change or deprecation?

No.

chrisohaver commented on plugin/cache/README.md:
```diff
@@ -57,7 +57,10 @@ cache [TTL] [ZONES...] {
   entry to the client. The responses have a TTL of 0. **DURATION** is how far back to consider
-  stale responses as fresh. The default duration is 1h.
+  stale responses as fresh. The default duration is 1h. **REFRESH_MODE** controls whether the attempt to refresh
+  the cache happens before (`immediate`) or after (`verified`) returning a response. This defaults to `immediate`.
```
> refresh the cache happens before (`immediate`) or after (`verified`) returning a response

"immediate" and "verified" are swapped here, resulting in incorrect description.
And the phrasing is confusing, probably better described using the option terminology.

* `verified` will first _verify_ that an entry is still unavailable from the source before sending the stale response to the client.
* `immediate` will _immediately_ send a response to the client before checking to see if the entry is available from the source.

atollena replied:
fixed, thank you for the suggestion.

chrisohaver commented on plugin/cache/handler.go:
```diff
@@ -38,9 +38,13 @@ func (c *Cache) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg)
+		rcode, err := c.doRefresh(ctx, state, crr)
+		if i == nil || (rcode == dns.RcodeSuccess || rcode == dns.RcodeNameError){
+			return rcode, err
+		}
```
Calling `c.doRefresh` generally results in sending a response to the client if a response is received - no matter what the response code is. As proposed, if the response code is not NXDOMAIN or NOERROR, the client can be sent two responses, the first one being the current upstream response, and the second one (which would be ignored by the client) being the stale cached entry.

atollena replied:
OK, I realise that I did not understand your comment in https://github.com/coredns/coredns/pull/4756#discussion_r672453981.

I wrapped the response writer as you suggested, so that it only writes if the response is nxdomain or noerror and exposes whether it wrote. I'm not quite sure that the behaviour is correct in case the underlying `ResponseWriter` returns an error to `WriteMsg`. As proposed the code just ignores the error and proceeds with serving stale entry -- my concern is whether this could in certain cases lead to double writing.

chrisohaver commented on plugin/cache/README.md:
```diff
@@ -57,7 +57,10 @@ cache [TTL] [ZONES...] {
+  stale responses as fresh. The default duration is 1h. **REFRESH_MODE** controls when the attempt to refresh
+  the cache happens. `verified` will first verify that an entry is still unavailable from the source before sending
+  the stale response to the client. `immediate` will immediately send the expired response to the client before 
+  checking to see if the entry is available from the source.  **REFRESH_MODE** defaults to `immediate`.
```
Please add a warning about that latency will be higher when using `verified`.  It was present in an earlier commit, but looks like it was removed in this last commit.

atollena replied:
I restored the warning (I changed a few words but kept the content).

chrisohaver commented on plugin/cache/cache.go:
```diff
@@ -225,6 +227,34 @@ func (w *ResponseWriter) Write(buf []byte) (int, error) {
+		w.refreshed = true
+		return w.ResponseWriter.WriteMsg(res) // stores to the cache and send to client
+	}
+	return nil // else discard
```
This should check the response code of the message directly (`res.Rcode == dns.RcodeSuccess || res.Rcode == dns.RcodeNameError`), rather than running the message through `Typify` which returns pseudo-codes (like `NoData`, which has a `NOERROR` return code).

atollena replied:
Done.

chrisohaver commented on plugin/cache/README.md:
```diff
@@ -57,7 +57,12 @@ cache [TTL] [ZONES...] {
+  the stale response to the client. `immediate` will immediately send the expired response to the client before 
+  checking to see if the entry is available from the source.  **REFRESH_MODE** defaults to `immediate`. Setting this
+  value to `verified` can lead to increased latency when serving stale responses for unavailable upstreams, but can
+  reduce the number of stale entries served, especially when resolution is infrequent.
```
Latency is always increased for serving stale entries with `verified` ...  not just for unavailable upstreams/sources.  

"can reduce" -> "will prevent": `verified` prevents stale entries from ever being served if an updated response can be retrieved from the source. 

```suggestion
  value to `verified` can lead to increased latency when serving stale responses, but will
  prevent stale entries from ever being served if an updated response can be retrieved from the source.
```

> especially when resolution is infrequent

I think better to remove this qualification. `verified` always prevents this, regardless of resolution frequency.  Conversely, `immediate` will serve  1 extra stale result to a client before updating the cache.  If there is high concurrency on the cache item, then many of those stale responses can be written to clients before the cache is updated.

atollena replied:
I took your suggestion.

chrisohaver reviewed (changes requested):
Code LGTM.  Just some minor tweaks to the readme now.

atollena committed:
plugin/cache: Add refresh mode setting to serve_stale

This PR adds an optional REFRESH_MODE parameter on the serve_stale configuration directive of the
cache plugin, which verifies that the upstream is still unavailable before returning stale entries.

Signed-off-by: Antoine Tollenaere <atollena@gmail.com>

lobshunter commented on plugin/cache/README.md:
```diff
@@ -57,7 +57,12 @@ cache [TTL] [ZONES...] {
   entry to the client. The responses have a TTL of 0. **DURATION** is how far back to consider
-  stale responses as fresh. The default duration is 1h.
+  stale responses as fresh. The default duration is 1h. **REFRESH_MODE** controls when the attempt to refresh
+  the cache happens. `verified` will first verify that an entry is still unavailable from the source before sending
```
A minor typo here, `verified` will cause a invalid value error

```suggestion
  the cache happens. `verify` will first verify that an entry is still unavailable from the source before sending
```

chrisohaver replied:
@lobshunter, This PR merged a couple months ago, and the typo has since been fixed.

lobshunter replied:
My bad, I checked out to this PR to see how it works, didn't notice the new fix. 😂
